### PR TITLE
feat(optimizer): support m <= rank AND rank <= n

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/5692

Mainly works for `ROW_NUMBER()`. 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs [hunk 1]
```diff
@@ -80,49 +80,77 @@ impl Rule for OverAggToTopNRule {
...
+                    return None;
+                }
+            }
+
+            if eq.len() > 1
+                || lb.len() > 1
+                || ub.len() > 1
+                || (!eq.is_empty() && (!lb.is_empty() || !ub.is_empty()))
+            {
+                tracing::error!("Failed to optimize multiple complex rank predicates");
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/rule/over_agg_to_topn.rs:130]: Just curious: why not just merging them together? For example, if `ub = [10, 20]`, then use `ub = 10`

**Review by chenzl25 — APPROVED** [review]
LGTM. BTW, `between` is also some kind of range.

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs [hunk 2]
```diff
@@ -80,49 +80,77 @@ impl Rule for OverAggToTopNRule {
...
-                    .ok()?
-                    .eval_row_const()
-                    .ok()??;
-                let v = *v.as_int64();
-                if v == 1 {
+            if !eq.is_empty() {
+                if eq[0] == 1 {
                     (1, 0)
                 } else {
+                    tracing::error!("Failed to optimize complex rank predicate rank={}", eq[0]);
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/rule/over_agg_to_topn.rs:137]: ditto.

File: src/frontend/planner_test/tests/testdata/insert.yaml [hunk 3]
```diff
@@ -32,32 +32,32 @@
   sql: |
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/insert.yaml:32]: Are there any e2e tests for this feature?

**Review by fuyufjh — COMMENTED** [review]
LGTM for rest.

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs (hunk 1 above)
Comment (xxchan) [reply src/frontend/src/optimizer/rule/over_agg_to_topn.rs:130]: Just KISS. But yes it's not hard, so fixed now.

File: src/frontend/planner_test/tests/testdata/insert.yaml (hunk 3 above)
Comment (xxchan) [reply src/frontend/planner_test/tests/testdata/insert.yaml:32]: added

**Comment by xxchan** [comment]
> `between` is also some kind of range.

`between` is bound to `<= AND >=`, so it's already supported. Added a test case.

**Pull request merged**